Initial draft of Covid-19 blog post (#229)
kiegroup/optaplanner-website: blog/2020-05-08-AIVersusCovid-19.adoc

= AI versus Covid-19: How Java helps nurses and doctors in this fight
:page-interpolate: true
:awestruct-author: Christopher-Chianelli
:awestruct-layout: blogPostBase

> :awestruct-tags: [use case]

Reviewer: Add a share_image_filename line like in the other blog posts. It goes into the the html head as a meta info so when social media show that url, it show that picture along.
So that share_image should be something that attracks, for example a screenshot or so.

Scheduling a hospital is tough. To maintain employees' happiness and health,
you need to take into account sick days, days off requests, and assure
the employees have ample time between shifts. This is doubly so in a pandemic,
where you need to minimize exposure to the new disease and have nurses with

> different risks. Let's take a look at how we can quickly and efficiently

Reviewer 2: > ... you need to ... have nurses with different risks.

I don't understand this. Risk of what? Did you mean nurses with different skills? That would make sense to me.

Author: Ex: older nurses, nurses with diabetes, etc. Or worded differently: Different susceptibility to COVID-19.

its ward (for example, we need a repository specialist nurse in a Covid-19
ward, as well as a nurse and a doctor)

* Finally, there are _availabilities_, which describes when an employee is unable to
work, desires to work, or is requesting time off

The following class diagram shows the Java classes that represent this domain:

image::CovidEmployeeRosteringClassDiagram.png[Class Diagram]

== The constraints

OptaWeb Employee Rostering comes with a set of builtin
constraints, which you can add to, modify or remove according to your needs.
A few examples:

* An employee cannot work two shifts that overlap; after all, how do you expect an
employee to be in two places at the same time? As such, this is a hard constraint,
meaning a schedule is infeasible if broken.

* Whenever possible, an employee should not be working on a day they request off.
Since the schedule remains feasible an employee is working on a day they want off,
this is a soft constraint.

* The higher an employee's Covid-19 susceptibility is, the less they should
work in a Covid-19 ward. This is another soft constraint.

These, and the rest of the constraints, can be found in the

> https://github.com/kiegroup/optaweb-employee-rostering/blob/covid-19/optaweb-employee-rostering-backend/src/main/resources/org/optaweb/employeerostering/service/solver/employeeRosteringScoreRules.drl[Drools constraints

Reviewer 2: This is unsafe. What if we rename or delete the branch one day? What if the branch continues to evolve and we rename or move the DRL file? I recommend to reference a specific commit here to avoid a broken link in the future.

> file]. They can also be implemented in Java using constraint streams.

Reviewer 2: Very soft suggestion: _... in Java using OptaPlanner Constraint Stream API_ or something like that. Why? To avoid Java non-experts thinking: Wow, I didn't know Java had constraint streams! What's that?

== Using the application

OptaWeb Employee Rostering comes with both a backend and frontend. You could use the
bundled frontend or use the backend's REST endpoints. The application by default
starts with demo data, allowing you to try out the application. Let's use one of
example tenant and solve its roster. Navigate to the "Scheduling" tab and click
the (?) button next to the score. You will see a detailed summary of all the constraints.

> In it, we see the constraint "Assign every shift" has 1116 matches, meaning we have

Reviewer 3: ```suggestion
On the bottom of the list, we see the constraint "Assign every shift" has 1116 matches, meaning we have
```

1116 unassigned shifts. Let's fix that right now.

image::ConstraintSummary1.png[Unassigned Shifts]

> Close the summary popup and click on the "Solve" button. OptaPlanner will begin solving

Reviewer 3: ```suggestion
Close the summary popup and click on the "Schedule" button. OptaPlanner will begin solving
```

the roster. After about 30 seconds, you'll see there are no more unassigned shifts --
that is, OptaPlanner assigned an employee to each of the 1116 shifts, all the while

> making sure no employee get two shifts at the same time, each employee has the required

Reviewer 2: ```suggestion
making sure no employee gets two shifts at the same time, each employee has the required
```
I think.

skills for their shift, and each employee has enough rest between shifts. Imagine
making a feasible schedule by assigning 1116 shifts by hand; it would take a good
while longer than 30 seconds!

image::ConstraintSummary1.png[Assigned Shifts]

While solving, OptaPlanner will keep looking for better solutions, gradually improving
the score. You might have already seen the schedule change several times already.
Let's stop solving for now. Click the Ward selector (left side of the toolbar) and
change to the ward "COVID-19 Ward 1". Every shift has a small circle at the top of
it denoting the Covid-19 susceptibility of its employee. A shield represents
inoculated employees; people who have acquired immunity to Covid-19. A near empty

> circle mean the employee has a low Covid-19 susceptibility, while a more filled in

Reviewer 2: ```suggestion
circle means the employee has a low Covid-19 susceptibility, while a more filled in
```

> circle mean a higher Covid-19 susceptibility. Finally, a red circle mean the employee

Reviewer 2: ```suggestion
circle means a higher Covid-19 susceptibility. Finally, a red circle means the employee
```

> is extremely susceptible to Covid-19, and should not placed in a ward with Covid-19.

Reviewer 2: ```suggestion
is extremely susceptible to Covid-19, and should not be placed in a ward with Covid-19.
```

Notice how there are no Extreme risk employees in the Covid-19 ward? Moreover, most
of the Covid-19's ward shifts either have inoculated or low risk employees.

image::CovidWardNoExtremeRisk.png[No Extreme Risk Employees in Covid-19 Ward]

One second, I am getting a call. Okay, I'm back. Beth P. Watt caught the flu and

> as such is not able to make their shift this Saturday. Luckily for us, OptaPlanner

Reviewer 2: ```suggestion
as such is not able to make her shift this Saturday. Luckily for us, OptaPlanner
```

Soft suggestion. AFAIK "their" can be used if the gender of the subject is uncertain, e.g. when speaking about "the user". In this sentence it sounds a bit off since Beth is clearly a female name.

schedule by assigning Beth P. Watt's two shifts to another employee. Also, note
that OptaPlanner changed only those two shifts; OptaPlanner will try to minimize
the number of published shifts changed (after all, nurses and doctors have a
life, which they plan around their schedules).

image::ScheduleAfterReplanning.png[Only the two shifts Beth P. Watt was unavailable for changed]

A few other things you can do with this application:

* You can "pin" an employee to a particular shift, meaning OptaPlanner
will not change that shift's employee when solving.

* You can create a rotation for the employees, giving them more consistency
in their day to day lives and allowing automatic provisioning of new shifts.

== Conclusion

As you just saw, OptaWeb Employee Rostering allows you to easily create
and optimize rosters, that does not only minimize Covid-19's effect on
health care workers, but also maximizes their happiness. This was only
possible as OptaPlanner has such a flexible and easy way to add new constraints,
allowing us to quickly modify OptaWeb Employee Rostering to accommodate
the new constraints that came with Covid-19. You might also be interested

> in https://developers.redhat.com/devnation/tech-talks/AI-vs-covid19-java/[Geoffery De Smet's talk on the application].

Reviewer: Also add a link to the source on github (correct branch) in the conclusion, togher with something "follow the readme to build and run it locally"

Author: I put the repo link in the "Using the application" section, since I am basically giving users directions on how to use the application, which only make sense if they could build/run it first. I might add it again in the conclusion.

> in https://developers.redhat.com/devnation/tech-talks/AI-vs-covid19-java/[Geoffery De Smet's talk on the application].

Reviewer 2: ```suggestion
in https://developers.redhat.com/devnation/tech-talks/AI-vs-covid19-java/[Geoffrey De Smet's talk on the application].
```
